pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/vf/rule/security/internal/ElEscapeDetector.java

```diff
@@ -28,22 +29,234 @@
 
 public final class ElEscapeDetector {
 
-    private static final Set<String> SAFE_EXPRESSIONS = new HashSet<>(Arrays.asList("id", "size", "caseNumber"));
-    private static final Set<String> NON_EMPTY_ARG_SAFE_RESOURCE = new HashSet<>(Arrays.asList("urlfor", "casesafeid",
-            "begins", "contains", "len", "getrecordids", "linkto", "sqrt", "round", "mod", "log", "ln", "exp", "abs",
-            "floor", "ceiling", "nullvalue", "isnumber", "isnull", "isnew", "isblank", "isclone", "year", "month",
```

Comment:
There are now two functions/properties missing: `linkto` and `nullvalue`. Is this correct, that they now create a violation if used without escaping? Previously we ignored these....

Reply:
NULLVALUE() accepts a test parameter and a substitute parameter. If the test parameter is null, it is replaced with the substitute parameter. i.e., `NULLVALUE(x, y)` is equivalent to `IF(ISNULL(x), y, x)`. Therefore, escaping is required.
LINKTO() *might* be safe, but I'm sincerely unsure. That being the case, I erred on the side of caution.

Reply:
I've spent some time messing around with LINKTO trying to get an XSS to work, and I can't seem to manage that. So I'm going to conclude that the method is safe and put LINKTO back in the list. Good catch.